lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,456 @@
+    def str_dotted_access(obj, name)
+      name.split('.').each do |part|
+        obj = obj.send(part)
+      end
+      obj.nil? ? nil : obj.to_s
+    end
+
+    # Computes a bundle id from the discriminator fields of `obj`.
+    #
+    # +discriminator_fields+ may include '.' as a separator, which is used to
+    # indicate object traversal.  This is meant to allow fields in the
+    # computed bundle_id.
+    # the return is an array computed by going through the discriminator fields
+    # in order and obtaining the str(value) object field (or nested object
+    # field) if any discriminator field cannot be found, ValueError is raised.
+    #
+    # @param obj [Object] an object.
+    # @param discriminator_fields [Array<String>] a list of discriminator
+    #     fields in the order to be to be used in the id.
+    # @raise [NoMethodError] if the object does not contain the named attribute.
+    # @return [Array<Object>] array of objects computed as described above.
+    def compute_bundle_id(obj, discriminator_fields)
+      result = []
+      discriminator_fields.each do |field|
+        result.push(str_dotted_access(obj, field))
+      end
+      result
+    end
+
+    # rubocop:disable Metrics/ClassLength
+
+    # Coordinates the execution of a single bundle.
+    #
+    # @attribute [r] bundle_id
+    #   @return [String] the id of this bundle.
+    # @attribute [r] bundled_field
+    #   @return [String] the field used to create the bundled request.
+    # @attribute [r] subresponse_field
+    #   @return [String] tptional field used to demultiplex responses.
+    class Task
+      attr_reader :bundle_id, :bundled_field,
+                  :subresponse_field
+
+      # rubocop:disable Metrics/ParameterLists
+
+      # @param api_call [Proc] used to make an api call when the task is run.
+      # @param bundle_id [String] the id of this bundle.
+      # @param bundled_field [String] the field used to create the
+      #     bundled request.
+      # @param bundling_request [Object] the request to pass as the arg to
+      #     the api_call.
+      # @param kwargs [Hash] the keyword arguments passed to the api_call.
+      # @param subresponse_field [String] optional field used to demultiplex
+      #     responses.
+      def initialize(api_call,
+                     bundle_id,
+                     bundled_field,
+                     bundling_request,
+                     kwargs,
+                     subresponse_field: nil)
+        @api_call = api_call
+        @bundle_id = bundle_id
+        @bundled_field = bundled_field
+        @bundling_request = bundling_request
+        @kwargs = kwargs
+        @subresponse_field = subresponse_field
+        @in_deque = []
+        @event_deque = []
+      end
+
+      # The number of bundled elements in the repeated field.
+      # @return [Numeric]
+      def element_count
+        sum = 0
+        @in_deque.each { |elts| sum += elts.count }
+        sum
```

Comment:
FYI: you could write these three lines as `@in_deque.reduce(0) { |sum, elts| sum + elts.count }`